### PR TITLE
Avoid locks on highly parallel paths in VisualStudioWorkspaceImpl

### LINKED ISSUE #54135: [Performance] Large amount of contention in VisualStudioWorkspaceImpl.SetDependencyNodeTargetIdentifier (0.4% CPU of solution open)
**Version Used**: 
17.0.0 Preview 1.0 [31401.12.main]

[Trace](https://microsoft.sharepoint.com/:f:/t/VSIDEPerformance/Etm7IDAYTSpIn9CAQYfXgYUBuUMciQ4xUku3e1KWZtLspw?e=YOQ2sB) (1404 projects)

There's such a large amount of contention in SetDependencyNodeTargetIdentifier that overhead of calling Monitor.Enter itself (not the blocking wait) is 0.4% of all CPU.

image

### LINKED ISSUE #54137: [Performance] Large amount of contention in VisualStudioWorkspaceImpl.SetMaxLanguageVersion (0.5% CPU of solution open)
**Version Used**: 
17.0.0 Preview 1.0 [31401.12.main]

[Trace](https://microsoft.sharepoint.com/:f:/t/VSIDEPerformance/Etm7IDAYTSpIn9CAQYfXgYUBuUMciQ4xUku3e1KWZtLspw?e=YOQ2sB) (1404 projects)

There's such a large amount of contention in VisualStudioWorkspaceImpl.SetMaxLanguageVersion that overhead of calling Monitor.Enter itself (not the blocking wait) is 0.5% of all CPU:

image

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #54135 
Fixes #54137

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -1589,8 +1588,10 @@ protected internal override void OnProjectRemoved(ProjectId projectId)
 
                 _projectToHierarchyMap = _projectToHierarchyMap.Remove(projectId);
                 _projectToGuidMap = _projectToGuidMap.Remove(projectId);
-                _projectToMaxSupportedLangVersionMap.Remove(projectId);
-                _projectToDependencyNodeTargetIdentifier.Remove(projectId);
+                // _projectToMaxSupportedLangVersionMap needs to be updated with ImmutableInterlocked since it can be mutated outside the lock
+                ImmutableInterlocked.TryRemove<ProjectId, string?>(ref _projectToMaxSupportedLangVersionMap, projectId, out _);
+                // _projectToDependencyNodeTargetIdentifier needs to be updated with ImmutableInterlocked since it can be mutated outside the lock
+                ImmutableInterlocked.TryRemove(ref _projectToDependencyNodeTargetIdentifier, projectId, out _);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:1594]: should we be asserting the return value of this?

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm.  though i worry about costs of ImmutableDictionary being payed in other ways (i.e. expensive to iterate) :-/

**Comment by sharwell** [comment]
> though i worry about costs of ImmutableDictionary being payed in other ways (i.e. expensive to iterate)

For at least one of the two cases, we already have an immutable dictionary lookup on the same path as the new one.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:1594]: No, these will only have a value if the project system externally set the value after adding a project.

**Comment by CyrusNajmabadi** [comment]
>For at least one of the two cases, we already have an immutable dictionary lookup on the same path as the new one.

It's not necessarily lookup, but also things like iteration.  I haven't audited the use of this value.  And, even if i did, it's enormously hard to tell the final impact of something like this.  For example, we might then see "iterating this type is now x% of Y scenario, up from soemthing lower".  

But, like i said, i'm ok proceeding with this.  If we do run into an issue, we can then contemplate alternative approaches.

**Comment by sharwell** [comment]
@CyrusNajmabadi There is no iteration for either of these collections. Each one contains _exactly one_ call to `TryGetValue` in the code:

https://github.com/dotnet/roslyn/blob/2d36a9ff4810d74e07394b889f4aa439e4104255/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L428


https://github.com/dotnet/roslyn/blob/2d36a9ff4810d74e07394b889f4aa439e4104255/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L1310

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi There is no iteration for either of these collections. Each one contains exactly one call to TryGetValue in the code:

Again, i'm not saying that only iteration woudl be a problem.  I'm saying that this may add some cost we dont' realize, in excahnge for getting rid of the locking cost.  These scenarios are tough (for me) as i'm not sure how ot gauge if such changes will be a 100% win, or will be a win for one type of perf issue, but a potential substantive loss for another.

**Pull request merged**